### ISSUE TITLE
CORS issue on the frontend while trying to make a POST request.

### ISSUE THREAD

**Issue by AlkenD** [polterguy/magic#191]
**Describe the bug**
The docs state that it is possible to to use the magic/system/auth/register route to integrate the auth functionality on any frontend but while making a POST request it throws CORS issue

error message if needed
```
Access to XMLHttpRequest at 'http://localhost:4444/magic/system/auth/register' 
from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight 
request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' 
header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. 
The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.
```

**To Reproduce**
Steps to reproduce the behaviour:

Clone and run the repo to test a basic auth env to reproduce the issue.

[Github Repo](https://github.com/AlkenD/test-frontend.git)

**Expected behaviour**
Since this is the only route which can be accessed by anyone it should not throw this error as the magic config frontend url is empty and defaults to Allow All Origin and successfully create a user and send the conformation mail.

**Screenshots**
--Not Applicable--

**Desktop (please complete the following information):**
--Not Relevant--

**Smartphone (please complete the following information):**
--Not Relevant--

**Additional context**

**Comment by polterguy** [comment]
Did you try to explicitly add the frontend URL of `http://localhost:3000` to the following section ...?

```json
{
  "magic": {
    "frontend": {
      "urls": "http://localhost:3000"
    },
```

The above allows for supplying a comma separated list of frontend URLs. I can't remember, but I think you'll need the _"scheme"_ parts too (the `http://` or `https://` parts)

**Comment by AlkenD** [comment]
> 

Ahh it needed a restart to solve the problem. Btw is there a way to allow all origns aswell ?

**Comment by AlkenD** [comment]
btw i encountered one more issue after restarting the backend the frontend is unable to detect it is this also an issue ?

**Comment by polterguy** [comment]
> Btw is there a way to allow all origins as well

Yes, this is what it does by default - However, Google Chrome prevents the request for security reasons, which is kind of a drag for us, since one backend might in theory have an infinite amount of frontends. The idea from Chrome's perspective is to avoid sending the JWT token to some malicious site, injected somehow into the HTTP request object.

Personally I don't like CORS, and I think it's the _"wrong solution to the wrong problem"_, something that can be seen by the fact of that it's the server declaring the CORS policy, but the client that's enforcing it, resulting in a somewhat ridiculous security model where the client can choose to ignore it - But it is what it is I suspect ... :/

You might try to play with the `withCredentials` parts of your HTTP object in React, and _manually add_ the `Authorization` HTTP header instead, which should in theory allow you to ignore CORS ...

As to your second problem, open the Chrome developer's tools and let me know what it says ...

The simplest would be to __remove__ the `withCredentials` parts from your frontend React code I suspect, remove all CORS form config, and manually add the `Authorization` HTTP header somehow. The only place I have previously encountered Magic actually _needing_ `withCredentials` is as I am doing LDAP or some other _"fancy"_ stuff that shouldn't be necessary unless you want to build a _"domain SSO authentication thing"_ (I don't recommend it ...)

**Comment by AlkenD** [comment]
> 

that's this helped me a lot to understand on some of the aspects of what to take care of in the future as for the second problem here are the chrome logs

<img width="605" alt="Screenshot 2022-07-25 at 6 41 45 PM" src="https://user-images.githubusercontent.com/96910178/180785416-7c7ef9bb-09f0-4832-9dbd-7bb08ee82d66.png">

i presume that the dashboard faces a cors either cause of the specified frontend urls that i configured else something else

and thanks for helping me clear these issue and amazing work with this project as a frontend dev it is hard to force myself to learn the backend stuff dunno why 😂. Either way hope this project grows even more

**Comment by polterguy** [comment]
Thank you, I'll convey the nice words to the rest of our team, they'll appreciate it ^_^

Psst, so you're good now or ...?

**Comment by AlkenD** [comment]
Unfortunately the issue is not resolved. these was another response that i forgot to mention it.

<img width="906" alt="Screenshot 2022-07-25 at 6 54 47 PM" src="https://user-images.githubusercontent.com/96910178/180787943-f6ee6adc-7f6c-47d7-a844-5ce9b8c7fd56.png">

and since i have no way to change anything in the magic.config ill have to re deploy the container and then see if i have to add the dashboard url as well but this shouldn’t be the case as the dashboard is supposed to work with the backend without any configuration

**Comment by polterguy** [comment]
If you add _one_ CORS frontend, you'll have to add _all_ CORS frontend. If you've got your React frontend running at `http://localhost:3000` and your Magic Dashboard running on `http://localhost:5555` (Docker images), this results in the following config ...

```json
{
  "magic": {
    "frontend": {
      "urls": "http://localhost:3000, http://localhost:5555"
    },
```

The reason is that the default CORS setting is _"accept everything"_, which crashes with your `withCredentials` setting on your React HTTP object. This results in that you've got two options.

* Remove CORS entirely and remove the `withCredentials` property on your React HTTP request object
* Add _all_ frontends that are supposed to access your backend (through a browser) in your above `frontend:urls` configuration setting ...

Let me know if this fixes things for you ...

**Comment by AlkenD** [comment]
> If you add _one_ CORS frontend, you'll have to add _all_ CORS frontend. If you've got your React frontend running at `http://localhost:3000` and your Magic Dashboard running on `http://localhost:5555` (Docker images), this results in the following config ...
> 
> ```json
> {
>   "magic": {
>     "frontend": {
>       "urls": "http://localhost:3000, http://localhost:5555"
>     },
> ```
> 
> The reason is that the default CORS setting is _"accept everything"_, which crashes with your `withCredentials` setting on your React HTTP request object. This results in that you've got two options.
> 
> * Remove CORS entirely and remove the `withCredentials` property on your React HTTP request object
> * Add _all_ frontends that are supposed to access your backend (through a browser) in your above `frontend:urls` configuration setting ...
> 
> Let me know if this fixes things for you ...

Yup its working now thanks a lot for your help 🥳

**Comment by polterguy** [comment]
NP, we live to serve :)

Let us know if you've got other issues, they help us improve the platform. One thing this particular question made me started pondering, was how to dynamically apply a CORS profile, _without_ having to restart the backend (which isn't always a choice) ... :)

**Comment by polterguy** [comment]
> that would be awesome cause if the user forget's to add the dashboard url then he has no choice left other than re deploying the entire thing

I know, and it's a drag. I'm not sure if this is even possible, but I'll try to figure it out ;)
